src/compiler/parser.ts

```diff
@@ -2429,8 +2429,16 @@ namespace ts {
         function parseJSDocType(): TypeNode {
             scanner.setInJSDocType(true);
             const dotdotdot = parseOptionalToken(SyntaxKind.DotDotDotToken);
+            const moduleSpecifier = parseOptionalToken(SyntaxKind.ModuleKeyword);
             let type = parseTypeOrTypePredicate();
             scanner.setInJSDocType(false);
+            if (moduleSpecifier) {
+                const moduleTag = createNode(SyntaxKind.JSDocNamepathType, moduleSpecifier.pos) as JSDocNamepathType;
+                while (token() !== SyntaxKind.CloseBraceToken && token() !== SyntaxKind.EndOfFileToken) {

```

Comment:
I realised that parseJSDocType is also called from parseJSDocParameter, not just single type expressions. That means its right terminators include `,` and `)` See the `ParsingContext.JSDocParameters` entry in `isListTerminator`.

That also means that we need tests for jsdoc-style functions:

```js
/** @type {function(module:xxxxxxxxxx, module:xxxxxxxxxx)} */
```

Reply:
I'm not sure whether jsdoc.app mentions whether namepaths can be used in function types, but I think we should support it even if it does not.